libvast/src/data.cpp

```diff
@@ -120,112 +122,112 @@ bool is_complex(const data& x) {
 bool is_recursive(const data& x) {
   return caf::visit(detail::overload([](const auto&) { return false; },
                                      [](const list&) { return true; },
-                                     [](const map&) { return true; }),
+                                     [](const map&) { return true; },
+                                     [](const record&) { return true; }),
                     x);
 }
 
 bool is_container(const data& x) {
+  // TODO: should a record be considered as a container?
   return is_recursive(x);
 }
 
-const data* get(const list& xs, const offset& o) {
-  const list* x = &xs;
-  for (size_t i = 0; i < o.size(); ++i) {
-    auto& idx = o[i];
-    if (idx >= x->size())
-      return nullptr;
-    auto d = &(*x)[idx];
-    if (i + 1 == o.size())
-      return d;
-    x = caf::get_if<list>(d);
-    if (!x)
-      return nullptr;
+record flatten(const record& r) {
```

Comment:
That's a good question. I think at this point we are very hand-waivy in this regard. When it comes to controlling `std::vector<data>`, a user could create a pathologic file (e.g., Zeek log or other format) with bizarre columns where this path gets triggered. Otherwise a user cannot trigger this code path directly, i.e., not through writing weird queries.

I don't think we can rely on TCO here, even though it would be convenient. Should we rewrite it iteratively as of this PR?

Reply:
Wouldnt just writing a schema with a very deeply nested struct (and then adding an event of this type) be enough to trigger the overflow?

If rewriting iteratively is too much effort for now, a simple solution could also be to introduce a hard-coded max recursion depth.

Reply:
Okay, I'll add a `max_depth` parameter.

Reply:
Hmmm, I gave this a another look and I don't know whether I want to do this as part of this PR. I'd like to go over a bunch of other recursive functions as well and either (i) see if we can rewrite them iteratively first, and if not, (ii) rewrite the implementation with hard-coded max depth.

Since (ii) may result in larger API changes, as in this case for `flatten`, I'd like to avoid scope creep if possible.

Reply:
Feel free to drop the issue if it's more complicated than anticipated, but shouldnt something like this work as a minimal workaround?

```
record flatten(const record& r, size_t depth = 0) {
  record result;
  if (depth >= defaults::max_record_nesting depth) {
    VAST_WARNING_ANON("reached max nesting depth for a record");
    return result;
  }
  [...]
```